File: src/tb_cascade/derive.py
"""Phase 3 - Derived variables.

Computes the analysis-grade variables that sit on top of the validated raw
table (`schema.validate`'d output of `io.load_raw`) but are not themselves
part of the raw data: age and age band, dose-adherence ratio, cascade time
intervals, one boolean/categorical flag per cascade node (Descriptive Study
Plan Sec 7, Steps 2-8), and a right-censoring flag (Sec 6 item 6, Sec 10).

`build_analysis_table` assembles all of these alongside the raw columns
into a single analysis-ready DataFrame, and `persist_analysis_table` writes
it to `Data/processed/analysis_ready_<run_date>.parquet`. Per the
Implementation Plan, that file is the *only* input Phase 4 (cascade/trend
analytics) is allowed to read -- every derived column downstream traces
back to one auditable place.

Design notes that apply across this module:
- Every function is a pure function of `df`; nothing here mutates its
  input or depends on global state besides the constants defined below.
- Wherever a value can't be computed (a required raw field is missing,
  or -- for ratios/flags built from multiple columns -- the inputs are
  ambiguous), the result is `pd.NA`/`pd.NaT` rather than a guessed value.
  Analysts downstream should be able to tell "not applicable/unknown"
  apart from a real zero or "no".
- Boolean combinations (`|`) are done directly on pandas nullable
  `"boolean"` columns so Kleene three-valued logic applies automatically:
  e.g. `True | NA = True` (eligible, full stop) but `False | NA = NA`
  (still unresolved), which is the behaviour an epidemiologic cascade
  needs and plain Python `or`/`.fillna(False)` would get wrong.
"""

from __future__ import annotations

from pathlib import Path

import pandas as pd

from tb_cascade import qc
from tb_cascade.config import PROCESSED_DATA_DIR

# --- age_at_screening ---------------------------------------------------

#: Fixed-width epidemiologic age bands per the Implementation Plan
#: (Phase 3 item 1): 0-14, 15-24, ..., 65+. `pd.cut` bin edges are
#: upper-bound-inclusive, so -inf/inf are used as the outer edges and the
#: interior edges are the band boundaries themselves.
_AGE_BIN_EDGES: list[float] = [-float("inf"), 14, 24, 34, 44, 54, 64, float("inf")]
_AGE_BIN_LABELS: list[str] = [
    "0-14",
    "15-24",
    "25-34",
    "35-44",
    "45-54",
    "55-64",
    "65+",
]


def age_at_screening(df: pd.DataFrame) -> pd.DataFrame:
    """Compute age at screening (years) and its fixed-width band.

    `age_years` = (`DateScreening` - `BirthDate`).days / 365.25, matching
    the formula already used by `qc.check_age_range` (that function is a
    QC range check only; this is the analysis-grade column it points to).

    `age_band` buckets `age_years` into the bands in `_AGE_BIN_LABELS`.
    It is left `pd.NA` when age itself is missing, *and* when age is
    negative (`BirthDate` after `DateScreening` -- already an impossible
    value flagged by `check_age_range`; this function does not invent a
    band for a value it cannot place in any real age group). Implausibly
    *old* ages (e.g. >100, also flagged by `check_age_range`) are still
    banded into "65+" here, since unlike a negative age they are at least
    a coherent (if suspect) age.

    Returns a two-column DataFrame (`age_years`, `age_band`) indexed like
    `df`, ready to be concatenated onto it.
    """
    both_present = df["BirthDate"].notna() & df["DateScreening"].notna()

    age_years = pd.Series(pd.NA, index=df.index, dtype="Float64")
    raw_years = (df["DateScreening"] - df["BirthDate"]).dt.days / 365.25
    age_years.loc[both_present] = raw_years.loc[both_present]

    age_band = pd.Series(pd.NA, index=df.index, dtype="string")
    bandable = both_present & (age_years >= 0)
    bandable = bandable.fillna(False)
    cut = pd.cut(
        age_years.loc[bandable].astype("float64"),
        bins=_AGE_BIN_EDGES,
        labels=_AGE_BIN_LABELS,
        right=True,
    )
    age_band.loc[bandable] = cut.astype("string")

    return pd.DataFrame({"age_years": age_years, "age_band": age_band}, index=df.index)


# --- adherence_ratio ------------------------------------------------------


def adherence_ratio(df: pd.DataFrame) -> pd.Series:
    """Compute the dose-adherence ratio `DosesTaken / SchemaDoses`.

    Guarded against division by zero/null per the Implementation Plan
    (Phase 3 item 2): the ratio is `pd.NA` unless both `DosesTaken` and
    `SchemaDoses` are present and `SchemaDoses` is strictly positive.
    """
    computable = (
        df["SchemaDoses"].notna()
        & (df["SchemaDoses"] > 0).fillna(False)
        & df["DosesTaken"].notna()
    )

    ratio = pd.Series(pd.NA, index=df.index, dtype="Float64", name="adherence_ratio")
    ratio.loc[computable] = (
        df.loc[computable, "DosesTaken"] / df.loc[computable, "SchemaDoses"]
    ).astype("float64")
    return ratio


# --- time_intervals ---------------------------------------------------------

#: Short labels for the dates in `qc.DATE_ORDER_SEQUENCE`, used to name the
#: interval columns below (e.g. `days_full_eval_to_treatment_start`).
_INTERVAL_NODE_LABELS: dict[str, str] = {
    "DateScreening": "screening",
    "DateCompleteExaminationTB": "full_eval",
    "DatePrevTreatmentStart": "treatment_start",
    "DateTreatmentScheme": "treatment_scheme",
    "DateOutcome": "outcome",
}

#: Programmatic initiation-delay targets used for Step 4's target metric
#: (Descriptive Study Plan Sec 7 Step 4: proportion initiated within 30/60
#: days of the diagnostic evaluation that made them eligible).
_INITIATION_TARGET_DAYS: list[int] = [30, 60]


def time_intervals(df: pd.DataFrame) -> pd.DataFrame:
    """Compute cascade time intervals (days) between consecutive milestones.

    One `days_<earlier>_to_<later>` column per consecutive pair in
    `qc.DATE_ORDER_SEQUENCE` -- the same canonical date sequence
    `check_date_order`/`date_order_pair_breakdown` already validate, so
    the interval columns line up exactly with what QC calls a "pair" and
    nothing here can drift out of sync with that ordering.

    This always includes `days_full_eval_to_treatment_start`, the
    diagnosis-to-treatment-start interval the Implementation Plan calls
    out by name (`DatePrevTreatmentStart - DateCompleteExaminationTB`).
    From that interval, `initiated_within_30d` / `initiated_within_60d`
    are derived: `pd.NA` if the interval itself is unavailable, `pd.NA`
    (not `False`) if the interval is negative (a reversed-date QC
    violation -- not a legitimate "missed the target" case), else
    whether the delay was within 30/60 days respectively.

    An interval is only computed where both its endpoint dates are
    present; otherwise it is `pd.NA` (nullable `"Int64"` days).
    """
    columns: dict[str, pd.Series] = {}
    for earlier, later in zip(qc.DATE_ORDER_SEQUENCE[:-1], qc.DATE_ORDER_SEQUENCE[1:]):
        label = f"days_{_INTERVAL_NODE_LABELS[earlier]}_to_{_INTERVAL_NODE_LABELS[later]}"
        both_present = df[earlier].notna() & df[later].notna()
        delta = pd.Series(pd.NA, index=df.index, dtype="Int64")
        delta.loc[both_present] = (
            df.loc[both_present, later] - df.loc[both_present, earlier]
        ).dt.days
        columns[label] = delta

    delay = columns["days_full_eval_to_treatment_start"]
    on_time_computable = delay.notna() & (delay >= 0).fillna(False)
    for target in _INITIATION_TARGET_DAYS:
        flag = pd.Series(pd.NA, index=df.index, dtype="boolean")
        flag.loc[on_time_computable] = (delay.loc[on_time_computable] <= target).to_numpy()
        columns[f"initiated_within_{target}d"] = flag

    return pd.DataFrame(columns, index=df.index)


# --- cascade_flags ------------------------------------------------------------

#: FinalOutcome code -> label, straight from the data dictionary
#: (1=no TB, 2=TB developed, 3=Unknown, 4=Other).
_FINAL_OUTCOME_LABELS: dict[int, str] = {
    1: "no_tb",
    2: "tb_developed",
    3: "unknown",
    4: "other",
}


def _recode(series: pd.Series, mapping: dict) -> pd.Series:
    """Recode a coded column into string labels via `mapping`, keeping
    `pd.NA` for missing values and for any value not present in `mapping`.
    """
    result = pd.Series(pd.NA, index=series.index, dtype="string")
    has_value = series.notna()
    result.loc[has_value] = series.loc[has_value].map(mapping).astype("string")
    return result


def _single_label(df: pd.DataFrame, flag_to_label: dict[str, str]) -> pd.Series:
    """Recode a set of mutually-exclusive 0/1 flag columns into one string
    label per record: the label for whichever flag is the single `True`
    among the group. `pd.NA` if zero or more than one flag is `True` --
    i.e. not applicable / not yet resolved, or an existing
    mutual-exclusivity QC violation (see `qc.check_diagnosis_mutual_exclusivity`
    / `qc.check_outcome_mutual_exclusivity`). This function reports that
    ambiguity as missing rather than guessing which label is "right".
    """
    cols = list(flag_to_label)
    flags_true = df[cols].fillna(False)
    n_true = flags_true.sum(axis=1)

    result = pd.Series(pd.NA, index=df.index, dtype="string")
    for col, label in flag_to_label.items():
        single = (n_true == 1) & flags_true[col]
        result = result.mask(single, label)
    return result


def cascade_flags(df: pd.DataFrame) -> pd.DataFrame:
    """One boolean/categorical column per cascade node, Steps 2-8.

    Per the Implementation Plan (Phase 3 item 4), this turns the
    structural cascade logic already encoded in
    `qc.STRUCTURAL_MISSINGNESS_RULES` into named, analysis-ready columns,
    so cascade tables downstream are plain `groupby().mean()` calls
    instead of repeated boolean logic. Covers every node named in
    Descriptive Study Plan Sec 7 Steps 2-8:

    - Step 2 (screening cascade): `reached_screening`, `reached_suspected`,
      `diaskintest_positive`, `reached_full_eval`.
    - Step 3 (diagnostic outcomes, mutually exclusive among the fully
      evaluated): `confirmed_active_tb`, `has_lti`, `no_tb_no_lti`,
      `no_tb_lti_unknown`, plus `diagnosis_branch` (single label summarizing
      the four).
    - Step 4 (LTI preventive-treatment cascade): `eligible_for_lti_tx`
      (`LTI` or `PrevTreatmentRec`, per the Plan's own eligibility
      definition), `lti_recommended`, `lti_prescribed`, `lti_started`.
    - Step 6 (adherence/completion): `completed_or_finished`
      (`TreatmentCompleted` or `TreatmentFinished`), plus `outcome_branch`
      (single label across the seven mutually exclusive Step 6 outcome
      flags). Step 5 (`RegBq`/`RegMfx`) is a regimen attribute, not a
      progression node, so it is not duplicated here.
    - Step 7 (incentive uptake): `supp_screening_received`,
      `supp_50pc_received`, `supp_100pc_received`, `supp_1yr_received`
      (`Supp1yearGr23` or `Supp1yearGr1`, whichever applies to the
      record's treatment group).
    - Step 8 (follow-up/final outcome): `rescreened_1yr`, `no_tb_after_1yr`,
      `rescreened_24mo`, `no_tb_after_24mo`, plus `final_outcome_category`
      (`FinalOutcome` recoded to a label).

    All boolean columns reuse the raw nullable `"boolean"` flags directly
    (aliased or OR'd via Kleene logic), so a value of `pd.NA` here means
    exactly what it means on the underlying raw column: not yet known or
    not applicable, never a guessed `False`.
    """
    cols: dict[str, pd.Series] = {}

    # Step 2 - screening cascade.
    cols["reached_screening"] = df["Screening"]
    cols["reached_suspected"] = df["SuspectedTB"]
    cols["diaskintest_positive"] = df["DiaskintestPositive"]
    cols["reached_full_eval"] = df["CompleteExaminationTB"]

    # Step 3 - diagnostic branch (mutually exclusive among fully evaluated).
    cols["confirmed_active_tb"] = df["ConfirmedDiagnosisTB"]
    cols["has_lti"] = df["LTI"]
    cols["no_tb_no_lti"] = df["NoTBNoLTI"]
    cols["no_tb_lti_unknown"] = df["NoTBLTIunknown"]
    cols["diagnosis_branch"] = _single_label(
        df,
        {
            "ConfirmedDiagnosisTB": "confirmed_tb",
            "LTI": "lti",
            "NoTBNoLTI": "no_tb_no_lti",
            "NoTBLTIunknown": "no_tb_lti_unknown",
        },
    )

    # Step 4 - LTI preventive-treatment cascade.
    cols["eligible_for_lti_tx"] = df["LTI"] | df["PrevTreatmentRec"]
    cols["lti_recommended"] = df["PrevTreatmentRec"]
    cols["lti_prescribed"] = df["PrevTreatmentPresc"]
    cols["lti_started"] = df["PrevTreatmentStart"]

    # Step 6 - adherence and completion.
    cols["completed_or_finished"] = df["TreatmentCompleted"] | df["TreatmentFinished"]
    cols["outcome_branch"] = _single_label(
        df,
        {
            "TreatmentCompleted": "completed",
            "TreatmentFinished": "finished",
            "TBdeveloped": "tb_developed",
            "TreatmentStopedMed": "stopped_med",
            "TreatmetnNotFinished": "not_finished",
            "TreatmentContinue": "continuing",
            "OutcomeNotKnown": "unknown",
        },
    )

    # Step 7 - incentive payment uptake.
    cols["supp_screening_received"] = df["SuppScreening"]
    cols["supp_50pc_received"] = df["Supp50pc"]
    cols["supp_100pc_received"] = df["Supp100pc"]
    cols["supp_1yr_received"] = df["Supp1yearGr23"] | df["Supp1yearGr1"]

    # Step 8 - follow-up and final outcome.
    cols["rescreened_1yr"] = df["Screening_y"]
    cols["no_tb_after_1yr"] = df["NoTbAfter_y"]
    cols["rescreened_24mo"] = df["Screening_24"]
    cols["no_tb_after_24mo"] = df["NoTbAfter_24"]
    cols["final_outcome_category"] = _recode(df["FinalOutcome"], _FINAL_OUTCOME_LABELS)

    return pd.DataFrame(cols, index=df.index)


# --- censoring_flag -----------------------------------------------------------


def censoring_flag(
    df: pd.DataFrame, analysis_date, window_months: int = 12
) -> pd.Series:
    """Flag records that cannot yet have a mature outcome (right-censoring).

    Per Descriptive Study Plan Sec 6.6/Sec 10: data collection is ongoing,
    so records enrolled close to `analysis_date` have not had the full
    `window_months` (default 12, per the Implementation Plan's signature)
    to reach a final outcome. A record is `True` (censored) if *both*:

    1. `DateScreening` (enrollment) falls within `window_months` of
       `analysis_date` -- i.e. after `analysis_date - window_months`, and
    2. `FinalOutcome` is still missing.

    Condition 2 matters: a record enrolled recently but that already has
    a recorded `FinalOutcome` reached its outcome quickly and is not
    censored, regardless of how close to `analysis_date` it was enrolled.
    This is the standard survival-analysis meaning of "censored" (no
    event observed, not enough follow-up time) rather than a bare
    enrollment-date cutoff, and is what Phase 4's incidence-rate
    calculation (Sec 7 Step 8) needs to exclude/handle correctly.

    `pd.NA` if `DateScreening` itself is missing (enrollment date unknown
    -- cannot assess).
    """
    analysis_ts = pd.Timestamp(analysis_date)
    cutoff = analysis_ts - pd.DateOffset(months=window_months)

    enrollment = df["DateScreening"]
    has_date = enrollment.notna()

    flag = pd.Series(pd.NA, index=df.index, dtype="boolean", name="censored")
    recently_enrolled = enrollment.loc[has_date] > cutoff
    has_outcome = df.loc[has_date, "FinalOutcome"].notna()
    flag.loc[has_date] = (recently_enrolled & ~has_outcome).to_numpy()
    return flag


# --- build_analysis_table / persist_analysis_table ---------------------------


def build_analysis_table(
    df: pd.DataFrame, analysis_date, window_months: int = 12
) -> pd.DataFrame:
    """Assemble the single analysis-ready table (Implementation Plan item 6).

    Concatenates every derived-variable group from this module onto the
    raw columns of `df`: `age_at_screening` (age_years, age_band),
    `adherence_ratio`, `time_intervals`, `cascade_flags`, and
    `censoring_flag` (computed against `analysis_date`/`window_months`).

    This is meant to be the *only* table Phase 4 (cascade/trend analytics)
    reads from -- everything it needs should already be a column here,
    so no analysis code re-derives cascade logic independently.
    """
    age = age_at_screening(df)
    adherence = adherence_ratio(df)
    intervals = time_intervals(df)
    flags = cascade_flags(df)
    censored = censoring_flag(df, analysis_date, window_months=window_months)

    return pd.concat([df, age, adherence, intervals, flags, censored], axis=1)


def persist_analysis_table(df: pd.DataFrame, run_date: str) -> Path:
    """Write an immutable Parquet snapshot of the analysis-ready table.

    The output path is `Data/processed/analysis_ready_<run_date>.parquet`,
    mirroring `io.snapshot`'s naming convention for the raw snapshot.
    Requires `pyarrow` (declared in `pyproject.toml`).
    """
    PROCESSED_DATA_DIR.mkdir(parents=True, exist_ok=True)
    out_path = PROCESSED_DATA_DIR / f"analysis_ready_{run_date}.parquet"
    df.to_parquet(out_path, index=False)
    return out_path

File: r/R/derive.R
.DIAGNOSIS_BRANCH_COLUMNS <- c(
  ConfirmedDiagnosisTB = "active_tb",
  LTI = "lti",
  NoTBNoLTI = "no_tb_no_lti",
  NoTBLTIunknown = "lti_unknown"
)

.OUTCOME_BRANCH_COLUMNS <- c(
  TreatmentCompleted = "completed",
  TreatmentFinished = "finished",
  TBdeveloped = "tb_developed",
  TreatmentStopedMed = "stopped_medical",
  TreatmetnNotFinished = "not_finished",
  TreatmentContinue = "continuing",
  OutcomeNotKnown = "unknown"
)

.FINAL_OUTCOME_LABELS <- c(
  "1" = "no_tb", "2" = "tb_developed", "3" = "unknown", "4" = "other"
)

#' Age in completed years at screening, and standard 10-year age bands
#'
#' Age is computed from `DateScreening` minus `BirthDate` (not from the
#' analysis/run date), truncated to completed years. Implausible ages
#' (negative, or unparseable because either date is missing) are not
#' corrected here -- see [check_age_range()] for the QC-stage flag -- they
#' simply come out as `NA`/negative and are excluded from `age_band`.
#'
#' @param birth_date,screening_date Date vectors of equal length.
#' @return A list with `age` (numeric, years) and `age_band` (ordered
#'   factor in 10-year bands, `"100+"` for the open-ended top band).
#' @export
compute_age <- function(birth_date, screening_date) {
  age <- floor(as.numeric(screening_date - birth_date) / 365.25)

  band_breaks <- c(seq(0, 100, by = 10), Inf)
  band_labels <- c(paste0(seq(0, 90, by = 10), "-", seq(9, 99, by = 10)), "100+")
  age_band <- cut(age, breaks = band_breaks, labels = band_labels, right = FALSE, ordered_result = TRUE)

  list(age = age, age_band = age_band)
}

#' Adherence ratio (doses taken / doses scheduled)
#'
#' `NA` whenever `schema_doses` is missing or zero, since the ratio is
#' undefined (avoids a div-by-zero `Inf`/`NaN`).
#' @param doses_taken,schema_doses Numeric vectors of equal length.
#' @return A numeric vector.
#' @export
adherence_ratio <- function(doses_taken, schema_doses) {
  dplyr::if_else(
    !is.na(schema_doses) & schema_doses > 0 & !is.na(doses_taken),
    doses_taken / schema_doses,
    NA_real_
  )
}

#' Cascade milestone time intervals, in days
#'
#' Computes the four adjacent gaps in [DATE_ORDER_SEQUENCE] plus the total
#' screening-to-outcome span (used as person-time in Step 8's incidence
#' rate). Intervals are signed: a negative value reflects a date-order
#' reversal already surfaced by [check_date_order()] -- it is not corrected
#' or clipped here.
#' @param df A data frame with the columns in [DATE_ORDER_SEQUENCE].
#' @return A tibble with one row per record in `df`: `days_screening_to_eval`,
#'   `days_eval_to_tx_start`, `days_tx_start_to_scheme`,
#'   `days_scheme_to_outcome`, `days_screening_to_outcome`.
#' @export
time_intervals <- function(df) {
  tibble::tibble(
    days_screening_to_eval = as.numeric(df$DateCompleteExaminationTB - df$DateScreening),
    days_eval_to_tx_start = as.numeric(df$DatePrevTreatmentStart - df$DateCompleteExaminationTB),
    days_tx_start_to_scheme = as.numeric(df$DateTreatmentScheme - df$DatePrevTreatmentStart),
    days_scheme_to_outcome = as.numeric(df$DateOutcome - df$DateTreatmentScheme),
    days_screening_to_outcome = as.numeric(df$DateOutcome - df$DateScreening)
  )
}

#' Treatment-initiation-within-target flags
#'
#' For each target in [INITIATION_TARGET_DAYS], whether
#' `days_eval_to_tx_start` (the `DateCompleteExaminationTB` to
#' `DatePrevTreatmentStart` delay) falls within that many days. A negative
#' delay (date-order reversal) is treated as `NA` -- not `TRUE` -- since the
#' true delay cannot be determined from inconsistent dates.
#' @param days_eval_to_tx_start Numeric vector, as returned by
#'   [time_intervals()].
#' @return A tibble with one logical column per entry in
#'   [INITIATION_TARGET_DAYS], named `initiated_within_<n>d`.
#' @export
initiation_target_flags <- function(days_eval_to_tx_start) {
  flags <- purrr::map(INITIATION_TARGET_DAYS, function(target) {
    dplyr::if_else(
      is.na(days_eval_to_tx_start) | days_eval_to_tx_start < 0,
      NA,
      days_eval_to_tx_start <= target
    )
  })
  names(flags) <- paste0("initiated_within_", INITIATION_TARGET_DAYS, "d")
  tibble::as_tibble(flags)
}

#' Map a set of mutually-exclusive 0/1 columns to a single labelled factor
#'
#' @param df A data frame.
#' @param column_labels A named character vector: names are columns of
#'   `df`, values are the factor label to use when that column is `1`.
#' @return A factor, `NA` where none of the columns is `1`.
#' @keywords internal
.branch_factor <- function(df, column_labels) {
  cols <- names(column_labels)
  label <- rep(NA_character_, nrow(df))
  for (col in cols) {
    is_set <- !is.na(df[[col]]) & df[[col]] == 1L
    label[is_set] <- column_labels[[col]]
  }
  factor(label, levels = unname(column_labels))
}

#' Screening-to-outcome cascade boolean/categorical flags (Study Plan Steps 2-8)
#'
#' @param df A data frame as returned by [load_raw()].
#' @return A tibble with one row per record in `df` and the following
#'   columns: `reached_screening`, `reached_suspected`,
#'   `diaskintest_positive`, `reached_full_eval` (Step 2);
#'   `confirmed_active_tb`, `has_lti`, `no_tb_no_lti`, `no_tb_lti_unknown`,
#'   `diagnosis_branch` (Step 3); `eligible_for_lti_tx`, `lti_recommended`,
#'   `lti_prescribed`, `lti_started` (Step 4); `completed_or_finished`,
#'   `outcome_branch` (Step 6); `supp_screening_received`,
#'   `supp_50pc_received`, `supp_100pc_received`, `supp_1yr_received`
#'   (Step 7); `rescreened_1yr`, `no_tb_after_1yr`, `rescreened_24mo`,
#'   `no_tb_after_24mo`, `final_outcome_category` (Step 8).
#' @export
cascade_flags <- function(df) {
  tibble::tibble(
    reached_screening = df$Screening == 1L,
    reached_suspected = df$SuspectedTB == 1L,
    diaskintest_positive = df$DiaskintestPositive == 1L,
    reached_full_eval = df$CompleteExaminationTB == 1L,
    confirmed_active_tb = df$ConfirmedDiagnosisTB == 1L,
    has_lti = df$LTI == 1L,
    no_tb_no_lti = df$NoTBNoLTI == 1L,
    no_tb_lti_unknown = df$NoTBLTIunknown == 1L,
    diagnosis_branch = .branch_factor(df, .DIAGNOSIS_BRANCH_COLUMNS),
    eligible_for_lti_tx = (df$LTI == 1L) | (df$PrevTreatmentRec == 1L),
    lti_recommended = df$PrevTreatmentRec == 1L,
    lti_prescribed = df$PrevTreatmentPresc == 1L,
    lti_started = df$PrevTreatmentStart == 1L,
    completed_or_finished = (df$TreatmentCompleted == 1L) | (df$TreatmentFinished == 1L),
    outcome_branch = .branch_factor(df, .OUTCOME_BRANCH_COLUMNS),
    supp_screening_received = df$SuppScreening == 1L,
    supp_50pc_received = df$Supp50pc == 1L,
    supp_100pc_received = df$Supp100pc == 1L,
    supp_1yr_received = dplyr::case_when(
      df$TreatGroup == 1L ~ df$Supp1yearGr1 == 1L,
      df$TreatGroup %in% c(2L, 3L) ~ df$Supp1yearGr23 == 1L,
      .default = NA
    ),
    rescreened_1yr = df$Screening_y == 1L,
    no_tb_after_1yr = df$NoTbAfter_y == 1L,
    rescreened_24mo = df$Screening_24 == 1L,
    no_tb_after_24mo = df$NoTbAfter_24 == 1L,
    final_outcome_category = factor(
      unname(.FINAL_OUTCOME_LABELS[as.character(df$FinalOutcome)]),
      levels = unname(.FINAL_OUTCOME_LABELS)
    )
  )
}

#' Right-censoring flag for individuals with insufficient follow-up
#'
#' Flags records whose `DateScreening` falls within `window_months` months
#' of `analysis_date` -- not enough time may have elapsed to reach a mature
#' outcome (Descriptive Study Plan SS6, point 6).
#' @param df A data frame with a `DateScreening` column.
#' @param analysis_date The date the analysis is being run as of; defaults
#'   to today.
#' @param window_months Censoring window, in months; defaults to
#'   [CENSORING_WINDOW_MONTHS].
#' @return A logical vector.
#' @export
censoring_flag <- function(df, analysis_date = Sys.Date(), window_months = CENSORING_WINDOW_MONTHS) {
  cutoff <- lubridate::`%m-%`(analysis_date, months(window_months))
  !is.na(df$DateScreening) & df$DateScreening > cutoff
}

#' Build the full analysis-ready table
#'
#' Binds every Phase 3 derived variable onto the raw loaded table: age/age
#' band, adherence ratio, the five cascade time intervals, the
#' initiation-within-target flags, the Step 2-8 cascade flags, and the
#' censoring flag.
#' @param df A data frame as returned by [load_raw()].
#' @param analysis_date The date the analysis is being run as of; defaults
#'   to today. Passed through to [censoring_flag()].
#' @return A tibble: `df` plus all derived columns.
#' @export
build_analysis_table <- function(df, analysis_date = Sys.Date()) {
  age <- compute_age(df$BirthDate, df$DateScreening)
  intervals <- time_intervals(df)

  dplyr::bind_cols(
    df,
    age = age$age,
    age_band = age$age_band,
    adherence_ratio = adherence_ratio(df$DosesTaken, df$SchemaDoses),
    intervals,
    initiation_target_flags(intervals$days_eval_to_tx_start),
    cascade_flags(df),
    censored = censoring_flag(df, analysis_date = analysis_date)
  )
}

#' Write an immutable Parquet snapshot of the analysis-ready table
#'
#' @param table A data frame as returned by [build_analysis_table()].
#' @param run_date Date identifying this run; defaults to today.
#' @param dir Output directory; defaults to [processed_data_dir()].
#' @return The path the snapshot was written to, invisibly.
#' @export
persist_analysis_table <- function(table, run_date = Sys.Date(), dir = processed_data_dir()) {
  if (!dir.exists(dir)) {
    dir.create(dir, recursive = TRUE)
  }
  path <- file.path(dir, sprintf("analysis_ready_%s.parquet", format(run_date, "%Y-%m-%d")))
  arrow::write_parquet(table, path)
  invisible(path)
}
